mono/metadata/unity-memory-info.c

```diff
@@ -723,86 +107,33 @@ ReportGenericClasses(MonoGenericClass *genericClass, gpointer user_data)
 {
 	ClassReportContext *context = (ClassReportContext *)user_data;
 
-	if (genericClass->cached_class != NULL)
+	if (genericClass->cached_class != NULL && genericClass->cached_class->inited)
 		context->callback(genericClass->cached_class, context->user_data);
 }
 
-void
+static void
+ReportImageSetClasses(MonoImageSet *imageSet, void* user_data)
+{
+	if (imageSet->array_cache)
+		g_hash_table_foreach(imageSet->array_cache, ReportHashMapListClasses, user_data);
+
+	if (imageSet->szarray_cache)
+		g_hash_table_foreach(imageSet->szarray_cache, ReportHashMapClasses, user_data);
+
+	if (imageSet->ptr_cache)
+		g_hash_table_foreach(imageSet->ptr_cache, ReportHashMapClasses, user_data);
+}
+
+MONO_API void
 mono_unity_class_for_each(ClassReportFunc callback, void *user_data)
 {
 	ClassReportContext reportContext;
 	reportContext.callback = callback;
 	reportContext.user_data = user_data;
+	// Report all assembly classes and assembly specific arrays
 	mono_domain_assembly_foreach(mono_domain_get(), ReportClassesFromAssembly, &reportContext);
+	// Report all generic classes
 	mono_metadata_generic_class_foreach(ReportGenericClasses, &reportContext);
+	// Report all image set arrays
```

Comment:
Nit, this comment is wrong as is seems pointer types are also processed?

Additionally, image sets are what are processed for `mono_metadata_generic_class_foreach` call above as well. Not sure it's worth changing, but I think this could be unified in a single image set processing pass.

Reply:
> Not sure it's worth changing, but I think this could be unified in a single image set processing pass.

I've looked at `mono_metadata_generic_class_foreach` as mono standard api that reports generics only and which is better not to touch to avoid sideeffects.
But I can totally add `gclass_cache` processing to `ReportImageSetClasses`, so that we have one loop over imagesets
Let me know if you prefer the last option - I'm happy to do the change

Reply:
I say leave as is, and let's revisit if it's performance bottleneck or issue.